localstack/services/s3/v3/provider.py

```diff
@@ -0,0 +1,1231 @@
+            http_method="HEAD",
+        )
+
+        response = HeadObjectOutput(
+            AcceptRanges="bytes",
+            **s3_object.get_system_metadata_fields(),
+        )
+        if s3_object.user_metadata:
+            response["Metadata"] = s3_object.user_metadata
+
+        # TODO implements if_match if_modified_since if_none_match if_unmodified_since
+        if checksum_algorithm := s3_object.checksum_algorithm:
+            # this is a bug in AWS: it sets the content encoding header to an empty string (parity tested)
+            response["ContentEncoding"] = ""
+            if request.get("ChecksumMode", "").upper() == "ENABLED":
+                response[f"Checksum{checksum_algorithm.upper()}"] = checksum  # noqa
+
+        if range_header := request.get("Range"):
+            range_data = parse_range_header(range_header, s3_object.size)
+            response["ContentLength"] = range_data.content_length
+
+        if s3_object.parts:
+            response["PartsCount"] = len(s3_object.parts)
+
+        if s3_object.version_id:
+            response["VersionId"] = s3_object.version_id
+
+        if s3_object.website_redirect_location:
+            response["WebsiteRedirectLocation"] = s3_object.website_redirect_location
+
+        # TODO: missing return fields:
+        # Expiration: Optional[Expiration]
+        # Restore: Optional[Restore]
+        # ArchiveStatus: Optional[ArchiveStatus]
+
+        # RequestCharged: Optional[RequestCharged]
+        # ReplicationStatus: Optional[ReplicationStatus]
+        # ObjectLockMode: Optional[ObjectLockMode]
+        # ObjectLockRetainUntilDate: Optional[ObjectLockRetainUntilDate]
+        # ObjectLockLegalHoldStatus: Optional[ObjectLockLegalHoldStatus]
+
+        return response
+
+    def delete_object(
+        self,
+        context: RequestContext,
+        bucket: BucketName,
+        key: ObjectKey,
+        mfa: MFA = None,
+        version_id: ObjectVersionId = None,
+        request_payer: RequestPayer = None,
+        bypass_governance_retention: BypassGovernanceRetention = None,
+        expected_bucket_owner: AccountId = None,
+    ) -> DeleteObjectOutput:
+        # TODO: implement bypass_governance_retention, it is done in moto
+        store = self.get_store(context.account_id, context.region)
+        if not (s3_bucket := store.buckets.get(bucket)):
+            raise NoSuchBucket("The specified bucket does not exist", BucketName=bucket)
+
+        if s3_bucket.versioning_status is None:
+            if version_id and version_id != "null":
+                raise InvalidArgument(
+                    "Invalid version id specified",
+                    ArgumentName="versionId",
+                    ArgumentValue=version_id,
+                )
+
+            found_object = s3_bucket.objects.pop(key, None)
+            # TODO: RequestCharged
+            if found_object:
+                self._storage_backend.remove(bucket, found_object)
```

Comment:
q: are there any IO exceptions we should guard against? or maybe it's better to just pass them through as 500 internal server error

Reply:
good question, I have to say I'm not sure yet. I feel like 500 is fine for now